File: typescript/Behavioral/19-State/State/BaseJobState.ts
import type { IJobState } from "./IJobState";
import type { JobCard } from "../Context/JobCard";

// Default behaviour: reject the action. Concrete states
// override only the transitions valid from themselves.
export abstract class BaseJobState implements IJobState {
  abstract readonly name: string;

  assign(_job: JobCard, _handyman: string): void {
    this.reject("assign");
  }
  start(_job: JobCard): void {
    this.reject("start");
  }
  pause(_job: JobCard, _reason: string): void {
    this.reject("pause");
  }
  resume(_job: JobCard): void {
    this.reject("resume");
  }
  complete(_job: JobCard): void {
    this.reject("complete");
  }
  approve(_job: JobCard): void {
    this.reject("approve");
  }

  private reject(action: string): never {
    throw new Error(`Cannot '${action}' a job in '${this.name}' state.`);
  }
}

File: csharp/Behavioral/19-State/State/BaseJobState.cs
namespace State;

// Default behaviour: reject the action. Concrete states
// override only the transitions valid from themselves.
public abstract class BaseJobState : IJobState
{
    public abstract string Name { get; }

    public virtual void Assign(JobCard job, string handyman) => Reject("assign");
    public virtual void Start(JobCard job)                   => Reject("start");
    public virtual void Pause(JobCard job, string reason)    => Reject("pause");
    public virtual void Resume(JobCard job)                  => Reject("resume");
    public virtual void Complete(JobCard job)                => Reject("complete");
    public virtual void Approve(JobCard job)                 => Reject("approve");

    private void Reject(string action) =>
        throw new InvalidOperationException(
            $"Cannot '{action}' a job in '{Name}' state.");
}
